### PR TITLE
Enable retry for Parquet writes [databricks]

### PR DESCRIPTION
**Pull request by andygrove**
Closes https://github.com/NVIDIA/spark-rapids/issues/8028

The first commit enables retries for Parquet writes (just changes a boolean parameter). This caused test regressions because we were sometimes casting timestamp columns to a cudF type that is not supported in Spark, and this caused issues when spilling batches. The casts are now performed after batches are spilled.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala [hunk 1]
```diff
@@ -165,12 +170,14 @@ abstract class ColumnarOutputWriter(context: TaskAttemptContext,
       }
       val startTimestamp = System.nanoTime
       withResource(sb.getColumnarBatch()) { cb =>
-        RmmRapidsRetryIterator.withRestoreOnRetry(cr) {
-          withResource(new NvtxRange(s"GPU $rangeName write", NvtxColor.BLUE)) { _ =>
-            withResource(GpuColumnVector.from(cb)) { table =>
-              scanTableBeforeWrite(table)
-              anythingWritten = true
-              tableWriter.write(table)
+        withResource(transform(cb)) { transformed =>
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala:173]: would be good to add a TODO + follow on issue to clean this up. If I understand correctly, if we keep the `Table` around we won't need to `transform`.

Comment (andygrove) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala:177]: Just to clarify this, we do still need to perform the transform. Previously, it was performed before creating the spillable batch, and now it is performed after getting the batch from the spillable batch. If we had a `SpillableTable` we could go back to performing it earlier.

**Comment by pxLi** [comment]
CI failed https://github.com/NVIDIA/spark-rapids/issues/8235

**Comment by sameerz** [comment]
build

**Comment by revans2** [comment]
The latest CI failures do not appear to be related to #8235.  

```
2023-05-12T01:18:19.7703936Z [2023-05-12T00:35:43.384Z] [gw2] [ 32%] FAILED ../../src/main/python/hive_delimited_text_test.py::test_basic_hive_text_write[hive-delim-text/simple-boolean-values-StructType(List(StructField(number,BooleanType,true)))-{}-TableWriteMode.CTAS][IGNORE_ORDER({'local': True}), APPROXIMATE_FLOAT] 23/05/12 00:35:43 WARN HiveExternalCatalog: Couldn't find corresponding Hive SerDe for data source provider csv. Persisting data source table `default`.`tmp_table_gw4_1415522135_0` into Hive metastore in **** SQL specific format, which is NOT compatible with Hive.
```

This shows up for a lot of different test failures in ORC, HIVE, and others.

**Comment by andygrove** [comment]
> This shows up for a lot of different test failures in ORC, HIVE, and others.

Yes, I see this locally too. Root cause is `Close called too many times`. I am looking into it.

**Comment by andygrove** [comment]
@revans2 @abellina I pushed a commit to resolve the double close issue (https://github.com/NVIDIA/spark-rapids/pull/8243/commits/f90d375b75db9b282243ccabe5921fe39a586a6c), but I am not confident that this is correct. Could you take a look?

**Comment by andygrove** [comment]
build

**Comment by andygrove** [comment]
build

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala [hunk 2]
```diff
@@ -156,6 +156,11 @@ abstract class ColumnarOutputWriter(context: TaskAttemptContext,
     }
   }
 
+  /** Apply any necessary casts before writing batch out */
+  def transform(cb: ColumnarBatch): ColumnarBatch = {
+    cb
+  }
+
   private[this] def writeBatchWithRetry(batch: ColumnarBatch): Long = {
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala:164]: so `writeBatchWithRetry` and `writeBatchNoRetry` are being called from `writeBatch`. The contract is that `writeBatch` should close the batch.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala [hunk 3]
```diff
@@ -321,6 +320,14 @@ class GpuParquetWriter(
     }
   }
 
+  override def transform(batch: ColumnarBatch): ColumnarBatch = {
+    val transformedCols = GpuColumnVector.extractColumns(batch).safeMap { cv =>
+      new GpuColumnVector(cv.dataType, deepTransformColumn(cv.getBase, cv.dataType))
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuParquetFileFormat.scala:325]: `deepTransformColumn` needs some comments (not an issue in this pr). I assume with the name that the input is not closed. With this latest patch, I think we need to close `batch` after we close `deepTransformColumn` on all the columns.

**Review by revans2 — DISMISSED** [review]

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala [hunk 4]
```diff
@@ -187,10 +198,12 @@ abstract class ColumnarOutputWriter(context: TaskAttemptContext,
     try {
       val startTimestamp = System.nanoTime
       withResource(new NvtxRange(s"GPU $rangeName write", NvtxColor.BLUE)) { _ =>
-        withResource(GpuColumnVector.from(batch)) { table =>
-          scanTableBeforeWrite(table)
-          anythingWritten = true
-          tableWriter.write(table)
+        closeOnExcept(transform(batch)) { b =>
```
Comment (abellina) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/ColumnarOutputWriter.scala:201]: why do we need `closeOnExcept` here?

**Comment by andygrove** [comment]
The blossom build is failing with:

```
2023-05-16T20:44:03.4035745Z ******** JOB LOGS with sensitive data redacted ************** 
2023-05-16T20:44:03.4036050Z 
2023-05-16T20:44:03.4036140Z {
2023-05-16T20:44:03.4036452Z   "errors" : [ {
2023-05-16T20:44:03.4036756Z     "status" : 404,
2023-05-16T20:44:03.4037091Z     "message" : "File not found."
2023-05-16T20:44:03.4037484Z   } ]
2023-05-16T20:44:03.4037761Z }
2023-05-16T20:44:04.9285018Z Cleaning up orphan processes
```

**Comment by abellina** [comment]
> The blossom build is failing with:
> 
> ```
> 2023-05-16T20:44:03.4035745Z ******** JOB LOGS with sensitive data redacted ************** 
> 2023-05-16T20:44:03.4036050Z 
> 2023-05-16T20:44:03.4036140Z {
> 2023-05-16T20:44:03.4036452Z   "errors" : [ {
> 2023-05-16T20:44:03.4036756Z     "status" : 404,
> 2023-05-16T20:44:03.4037091Z     "message" : "File not found."
> 2023-05-16T20:44:03.4037484Z   } ]
> 2023-05-16T20:44:03.4037761Z }
> 2023-05-16T20:44:04.9285018Z Cleaning up orphan processes
> ```

I think that's because of recent link changes in CI. I'll send you a direct message, but here's what I see:

```
[2023-05-16T20:05:47.876Z] java.lang.IllegalStateException: Close called too many times ColumnVector{rows=512, type=LIST, nullCount=Optional.empty, offHeap=(ID: 8889240 0)}
[2023-05-16T20:05:47.876Z]  at ai.rapids.cudf.ColumnVector.close(ColumnVector.java:269)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.GpuColumnVector.close(GpuColumnVector.java:1124)
[2023-05-16T20:05:47.876Z]  at org.apache.spark.sql.vectorized.ColumnarBatch.close(ColumnarBatch.java:48)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.RapidsPluginImplicits$AutoCloseableColumn.safeClose(implicits.scala:56)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.Arm$.withResource(Arm.scala:31)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatchWithRetry$1(ColumnarOutputWriter.scala:172)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.ColumnarOutputWriter.$anonfun$writeBatchWithRetry$1$adapted(ColumnarOutputWriter.scala:166)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$AutoCloseableAttemptSpliterator.next(RmmRapidsRetryIterator.scala:424)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.next(RmmRapidsRetryIterator.scala:528)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryAutoCloseableIterator.next(RmmRapidsRetryIterator.scala:458)
[2023-05-16T20:05:47.876Z]  at scala.collection.Iterator.foreach(Iterator.scala:941)
[2023-05-16T20:05:47.876Z]  at scala.collection.Iterator.foreach$(Iterator.scala:941)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.foreach(RmmRapidsRetryIterator.scala:477)
[2023-05-16T20:05:47.876Z]  at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:162)
[2023-05-16T20:05:47.876Z]  at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:160)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.foldLeft(RmmRapidsRetryIterator.scala:477)
[2023-05-16T20:05:47.876Z]  at scala.collection.TraversableOnce.sum(TraversableOnce.scala:221)
[2023-05-16T20:05:47.876Z]  at scala.collection.TraversableOnce.sum$(TraversableOnce.scala:221)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.RmmRapidsRetryIterator$RmmRapidsRetryIterator.sum(RmmRapidsRetryIterator.scala:477)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.ColumnarOutputWriter.writeBatchWithRetry(ColumnarOutputWriter.scala:193)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.ColumnarOutputWriter.writeBatch(ColumnarOutputWriter.scala:153)
[2023-05-16T20:05:47.876Z]  at com.nvidia.spark.rapids.ColumnarOutputWriter.writeAndClose(ColumnarOutputWriter.scala:115)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.sql.rapids.GpuSingleDirectoryDataWriter.write(GpuFileFormatDataWriter.scala:173)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.sql.rapids.GpuFileFormatDataWriter.writeWithIterator(GpuFileFormatDataWriter.scala:87)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$executeTask$1(GpuFileFormatWriter.scala:338)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.util.Utils$.tryWithSafeFinallyAndFailureCallbacks(Utils.scala:1473)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.sql.rapids.GpuFileFormatWriter$.executeTask(GpuFileFormatWriter.scala:345)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.sql.rapids.GpuFileFormatWriter$.$anonfun$write$15(GpuFileFormatWriter.scala:264)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.scheduler.Task.run(Task.scala:131)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.executor.Executor$TaskRunner.$anonfun$run$3(Executor.scala:497)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1439)
[2023-05-16T20:05:47.877Z]  at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:500)
[2023-05-16T20:05:47.877Z]  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
[2023-05-16T20:05:47.877Z]  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
```

**Review by revans2 — APPROVED** [review]

**Comment by andygrove** [comment]
build

**Review by abellina — APPROVED** [review]

**Pull request merged**